### ISSUE TITLE
Websockets improvement

### ISSUE THREAD

**Issue by Jabbey92** [starlite-api/starlite#199]
While the Starlette implementation of websockets is well done I think it could use a decent wrapper. Something with some more ... batteries. 

My proposal is simple at the moment but I think it would fit very well, Something along the lines of how routing works already. Essentially designed to be used as follows

from starlite import WebSocket, websocket

class Item(BaseModel):
    name: str

```
@websocket(path="/ws_path/{data}", data=Item)
async def my_websocket_handler(socket: WebSocket, data) -> None:
    async for message in socket:
        message.data
```

I figured this as one of the ways of implementation, as using it as a type hint in the function declaration implies it is meant to be used as a variable and that could be counter-intuitive. Although I have thought of several other ways this may be done.

```
# parsing whatever is in the string as the type, although would not be very friendly to use

@websocket(path="/ws_path/{data: Item}")
async def my_websocket_handler(socket: WebSocket) -> None:
    ...

# passing the data in as a variable, although a custom class would need to be created to hint
# that it will be iterated over. As well as likely another version of `starlite.WebSocket` that does
# not get iterated over or ditch the current version and make the socket var passed to the 
# function no longer iterated over thus making in these examples (`data`) now what you iterate over

@websocket(path="/ws_path/{data}")
async def my_websocket_handler(socket: WebSocket, data: typing.AsyncIterable[str]) -> None:
    ...

# or

async def my_websocket_handler(socket: WebSocket, data: typing.AsyncIterable[str]) -> None:
    ...

# parsing whatever is in the string as the type, although would not be very friendly to use

@websocket(path="/ws_path/{data}")
async def my_websocket_handler(socket: WebSocket) -> None:
    ...

```

This is in no way a final decision, just a proposition. I do plan to supply a PR because I love this project and just want to have my hand at helping.

**Comment by Goldziher** [comment]
Hi @Jabbey92.

First thanks for wanting to contribute. 

As to what you are proposing above - I don't quite understand what does data do in `path="/ws_path/{data}"`, is this supposed to be a url parameter? 

We already parse and handle parameters (URL, query, header and cookie) for websockets - same as we do for REST http endpoints. See this part of the docs: https://starlite-api.github.io/starlite/usage/3-parameters/

as for using the `data` keyword in the websocket_handler function, i.e. 

```python
from typing import AsyncIterable

async def my_websocket_handler(socket: WebSocket, data: AsyncIterable[Item]) -> None:
```

Basically now the user needs to call `websocket.receive_json()` and then either use the dictionary as is, or pass it to a constructor. So this could be a substantial improvement.

You will need to extend the `starlite.connection.WebSocket` class, and also work on the kwargs injection etc.

**Comment by Jabbey92** [comment]
_My sincerest apologies, like I said in discord I was very distracted while writing this and never even referred to my code I had wrote about this a month ago._

Thank you for being understanding and not brushing me aside with such an off the wall issue. Allow me the chance to re-iterate my intentions, and ask some more appropriate quetions hopefully.

I want to start by re-defining my goal with this issue.

From a strictly objective approach, I think that this is a more intuitive way to use websockets.

```
from typing import Any, Dict
from starlite import WebSocket, websocket
from pydantic import BaseModel

class Item(BaseModel):
    name: str
    ...


@websocket(path="/ws_path")
async def my_websocket_handler(socket: WebSocket[Item, Dict[str, Any]]) -> None:
    async for item in socket:
        print(item.name)
        ...
        socket.send({"value": 3})
```

**Question 1**
Is it safe to assume that in most cases one ws endpoint will imply one data type transfer? In this case it would be receiving `Item` and sending a dict

**Question 2**
It doesn't seem necessary to make `websocket.acccept()` and `websocket.close` a requirement to explicitely use unless desired.

- Either build them into the `__aiter__`
- Or (probably more sound) make the websocket into a context manager, this way the `__aiter__` functionality is required per say

Unless I'm not thinking of anything.

**Comment by peterschutt** [comment]
+1 from me for the `async for` pattern.

Let me know if I'm following correctly (BTW, websockets aren't a space that I work in, so if I'm misunderstanding anything, LMK). 

The proposal is that `WebSocket` is generic on send and receive type? We'd validate and parse inbound data inside `__aiter__` using the type specified via the generic class parameter so that iterating over the socket yields instances of the parsed type?

What does specifying the send type give us?

I remember going down a rabbit hole to find a reliable way to get access to generic type parameter values at runtime for something but my context has switched so far away from that stuff recently that I can't remember the specifics. This is in my search history: https://stackoverflow.com/questions/48572831/how-to-access-the-type-arguments-of-typing-generic - maybe it's not an issue, but just springs to mind.

**Comment by Jabbey92** [comment]
@peterschutt 

Thank you for the consideration!

> What does specifying the send type give us?

I figure, if not for anything else, documentation purposes.

Also something I considered:

Another option for if multiple types are expected for one endpoint is to overload the functions. While I dont think a decorator style like I suggested earlier would work. But there are other ways to do this in pure python.

**Comment by Goldziher** [comment]
@Jabbey92, so the issue with what you propose is that our `WebSocket` class already accepts generics - WebSocket[User, Auth]` - same as our `Request` class. 

I think the simplest way is to use the `data` keyword for the expected data type, or maybe alternatively add new keywords?

**Comment by Jabbey92** [comment]
@Goldziher 

So you're saying that there would be no way to change this? Why must it accept those generics? Otherwise there would be no real way for function documentation.

Maybe it can be a kwarg in the decorator?

```py
@websocket(path="/ws_path", receives=Item, sends=Dict[str, Any])
async def my_websocket_handler(socket: WebSocket) -> None:
    async for item in socket:
        print(item.name)
        ...
```

Would this cause breaking changes?

Excuse me for any misunderstanding, I see this in the code, although I don't see it's use in the documentation.

The code get's a little hard to follow, where is it used? I looked briefly through both starlite and starlette code just not fully grasping it's purpose is all.

**Comment by Jabbey92** [comment]
Just thinking, we could also give `Websocket` the functionality of being iterable.

It would also have 2 new properties: `recieve_type` and `send_type` with default values on initial creation.

Then create a separate class. Say `ParsedWebsocket` that returns a `WebSocket` with those new properties filled in. Thus fulfilling the type hinting within the function declaration. Also with the added benefit of backwards compatibility.

Example usage:

```py
@websocket(path="/ws_path")
async def my_websocket_handler(socket: ParsedWebSocket[Item, Dict[str, Any]]) -> None:
    async for item in socket:
        print(item.name)
        ...
```

**Comment by peterschutt** [comment]
> where is it used? 

Do you mean the `User` and `Auth` generics to `WebSocket`?

The generics define the static type of the return values of the `WebSocket.{user,auth}` properties:

```python
# connection.py
class WebSocket(StarletteWebSocket, Generic[User, Auth]):
    @property
    def app(self) -> "Starlite":
        return cast("Starlite", self.scope["app"])

    @property
    def user(self) -> User:
        if "user" not in self.scope:
            raise ImproperlyConfiguredException("'user' is not defined in scope, install an AuthMiddleware to set it")
        return cast(User, self.scope["user"])

    @property
    def auth(self) -> Auth:
        if "auth" not in self.scope:
            raise ImproperlyConfiguredException("'auth' is not defined in scope, install an AuthMiddleware to set it")
        return cast(Auth, self.scope["auth"])

    @property
    def query_params(self) -> Dict[str, Any]:  # type: ignore[override]
        return parse_query_params(self)
```

Here's where I'm at with all this. Firstly, no one is arguing that there aren't improvements to be made with to the websockets interface. However, there's lots of ways to skin the cat, and the devil is always in the detail.

First, I think we should have a minimal example application that demonstrates the current websockets pattern of use. We can do this as another example application inside the `starlite-api` org here on gh. It should be simple, and complete - i.e., anyone can easily clone the repo and follow some basic instructions in the readme to get the app up and running. Then we can use that app to try out different proposals and see what works and doesn't. Would you be interested in contributing to that @Jabbey92?

**Comment by peterschutt** [comment]
See:

https://github.com/starlite-api/starlite-websockets-hello-world

There is nothing websockets about it yet, just a pure clone of the other "hello-world" app in the org.

**Comment by peterschutt** [comment]
https://gist.github.com/whitecoop/157702ab76fb8243720fb318abd85341

**Comment by Jabbey92** [comment]
@peterschutt 

> Would you be interested in contributing to that @Jabbey92?

Absolutely, I've forked it and working on a PR currently.
